Split L3 Cache Topology Awareness in CPU Manager

sphrasavath opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Enhancement Split L3 Cache Topology Awareness in CPU Manager
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes enhancements/enhancements
[KEP4800](https://github.com/kubernetes/enhancements/issues/4800)
#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/issues/4800
[Design doc]:  https://docs.google.com/document/d/1LpnMjGNsQyHOuVHMktIrjZsdRw9aKZ8djt354nAno6M/edit?usp=sharing
```

ffromani commented:
/ok-to-test

ffromani commented:
hi! thanks for the contribution! the policy ot the kubernetes project is to not accept merge commits in the PRs, please update when you get a chance

ffromani commented:
priority TBD depending on the KEP conversation

ffromani commented:
/kind feature

ffromani commented:
/cc

ffromani commented:
/triage accepted

because KEP pending

ffromani commented:
/priority important-longterm

as the issue https://github.com/kubernetes/enhancements/issues/2621 was tentatively included in the upcoming 1.32 planning. Re-prioritization may be in order depending on the community feedback.

sphrasavath committed:
Add takeByTopologyUnCoreCachePacked if policy option align-cpus-by-uncorecache is enabled.
 Adding new function to evaluate uncore cache id. Reverse allocation logic.
 Implement preferAlignByUncorecache within TakeByTopologyNUMAPacked, along with new test cases.

wongchar commented:
/test pull-kubernetes-e2e-gce

kannon92 commented:
/hold

pacoxu commented:
> /hold

/unhold 
as the KEP was merged https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/4800-cpumanager-split-uncorecache/README.md 



> #### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
> ```
> [KEP]: https://github.com/kubernetes/enhancements/issues/2621
> [Design doc]:  https://docs.google.com/document/d/1LpnMjGNsQyHOuVHMktIrjZsdRw9aKZ8djt354nAno6M/edit?usp=sharing
> ```

@sphrasavath would you update the PR description to the new KEP id and link?

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -118,6 +120,23 @@ func (n *numaFirst) takeFullSecondLevel() {
 	n.acc.takeFullSockets()
 }
 
+// In Split L3 Topology, we take from the sets of uncorecache as the third level
```
let's make a pass to make sure the terminology in the implementation PR match what we used in the KEP, the option name is the prime example but we should also review therms like uncore, L3/LLC and so forth.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -184,6 +207,18 @@ func (s *socketsFirst) sortAvailableSockets() []int {
+// sit directly below NUMA Nodes in the memory hierchy
+func (s *socketsFirst) sortAvailableUnCoreCaches() []int {
+	var result []int
+	for _, uncore := range s.acc.sortAvailableNUMANodes() {
```
why `uncore`?
In addition, this logic seems identical to the `numaFirst` case, is that intentional?

ajcaldelas replied:
Yes, this is intentional, we have a couple methods here defined that follow the same logic

ffromani replied:
thanks @wongchar, will review again more carefully.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -346,6 +387,24 @@ func (a *cpuAccumulator) freeSockets() []int {
+
+// Returns all UnCoreCache IDs as a slice sorted by sortAvailableUnCoreCache().
+func (a *cpuAccumulator) allUnCoreCache() []int {
+	all := []int{}
+	all = append(all, a.numaOrSocketsFirst.sortAvailableUnCoreCaches()...)
+	return all
+}
+
```
why do we copy the returned slice?
can we just
```
return a.numaOrSocketsFirst.sortAvailableUnCoreCaches()
```
?

ajcaldelas replied:
I'm gonna agree with this, yeah we can do this in one line

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +577,45 @@ func (a *cpuAccumulator) takeFullSockets() {
+		if !a.needsAtLeast(cpusInUnCore.Size()) {
+			continue
+		}
+		klog.V(4).InfoS("takeFullCCD: claiming CCD", "uncore", uncore)
```
let's add consistent logging to what other functions have, and IIRC we don't add logging here

ajcaldelas replied:
Is there specific docs that talk about where we are suppose to log? I'm seeing it happen in the other similar spots this looks consistent

ffromani replied:
ok, let's postpone this for now

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +577,45 @@ func (a *cpuAccumulator) takeFullSockets() {
+// First try to take partial uncorecache (CCD), if available and the request size can fit w/in the uncorecache.
+// Second try to take the full CCD if available and need is at least the size of the uncorecache group.
+func (a *cpuAccumulator) takeUnCoreCache() {
+	// check if SMT ON
```
stale comment?

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +577,45 @@ func (a *cpuAccumulator) takeFullSockets() {
+	// check if SMT ON
+
+	for _, uncore := range a.allUnCoreCache() {
+		numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore() // this is another new change
```
I don't get the comment here

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +577,45 @@ func (a *cpuAccumulator) takeFullSockets() {
+	for _, uncore := range a.allUnCoreCache() {
+		numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore() // this is another new change
+
+		var freeCPUsInUncorecache cpuset.CPUSet
+		// need to get needed cores in uncorecache
+		freeCoresInUncorecache := a.details.CoresNeededInUnCoreCache(numCoresNeeded, uncore)
+		klog.V(2).InfoS("free cores from a.details list: ", "freeCoresInUncorecache", freeCoresInUncorecache)
+		for _, coreID := range freeCoresInUncorecache.List() {
+			freeCPUsInUncorecache = freeCPUsInUncorecache.Union(a.topo.CPUDetails.CPUsInCores(coreID))
+		}
+		klog.V(2).InfoS("freeCPUsInUncorecache  : ", "freeCPUsInUncorecache", freeCPUsInUncorecache)
+		if a.numCPUsNeeded == freeCPUsInUncorecache.Size() {
+			klog.V(4).InfoS("takePartialUncore: claiming cores from Uncorecache ID", "uncore", uncore)
+			a.take(freeCPUsInUncorecache)
+		}
+		// take full Uncorecache if the numCPUsNeeded is greater the L3 cache size
+		a.takeFullUnCore()
+
+		if a.isSatisfied() {
+			return
+		}
+	}
```
is this logic matching the intent described in the KEP? IIRC we should take full uncore group first, then try to allocate the reminder and the code as I read it seems to be doing the opposite

ffromani replied:
It seems it doesn't: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4800-cpumanager-split-uncorecache#design-details
```

    If the required CPUs are greater than or equal the size of the total amount of CPUs to an uncore cache, assign the full uncore cache CPUs if they are all unscheduled within the group. Subtract the amount of scheduled CPUs from the quantity of required CPUs.
        Continue scanning uncore caches in numerical order and assigning full uncore cache CPU groups until the required quantity is less than the total number of CPUs to a uncore cache group.
    If the required CPUs are less than the size of the total amount of CPUs to an uncore cache, scan through each uncore cache index in numerical order starting from the first index, and assign CPUs if there are enough available CPUs within the uncore cache group.
```

I'd expect to call first `takeFullUnCore()` then try to do partial allocation if cores are still needed

ffromani replied:
Overall, I finally realized that the code does what the KEP describes, which is good ofc. What we can and should do is to gather feedback from early user and review our algorithms first in the KEP and then in the code, tuning as needed. We can do this staying in alpha in 1.33 and postponing the beta graduation.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -702,6 +800,59 @@ func takeByTopologyNUMAPacked(topo *topology.CPUTopology, availableCPUs cpuset.C
 
+// takeByTopologyUnCoreCachePacked uses the "packed" sorting strategy similar to takeByTopologyNUMAPacked.
+// It includes an additional level of sorting by uncorecache
+func takeByTopologyUnCoreCachePacked(topo *topology.CPUTopology, availableCPUs cpuset.CPUSet, numCPUs int, cpuSortingStrategy CPUSortingStrategy) (cpuset.CPUSet, error) {
```
we should merge this logic in `takeByTopologyPacked` rather than duplicate the function

ffromani replied:
this module is not part of the kubernetes golang exported packages, usablie by third parties. And this function is private anyway :)

ffromani commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -33,6 +33,7 @@ const (
 	DistributeCPUsAcrossNUMAOption  string = "distribute-cpus-across-numa"
 	AlignBySocketOption             string = "align-by-socket"
 	DistributeCPUsAcrossCoresOption string = "distribute-cpus-across-cores"
+	AlignByUnCoreCacheOption        string = "align-cpus-by-uncorecache"
```
let's make the name consistent with the KEP

ffromani commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -43,6 +44,7 @@ var (
 	)
 	betaOptions = sets.New[string](
 		FullPCPUsOnlyOption,
+		AlignByUnCoreCacheOption,
```
we need to move this back to `alpha` options per the KEP

ffromani commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -138,6 +148,10 @@ func NewStaticPolicyOptions(policyOptions map[string]string) (StaticPolicyOption
 		return opts, fmt.Errorf("static policy options %s and %s can not be used at the same time", DistributeCPUsAcrossNUMAOption, DistributeCPUsAcrossCoresOption)
 	}
 
+	if opts.AlignByUnCoreCacheOption && opts.DistributeCPUsAcrossCores {
+		return opts, fmt.Errorf("static policy options %s and %s can not be used at the same time", AlignByUnCoreCacheOption, DistributeCPUsAcrossCoresOption)
+	}
+
```
we probably need also a similar check for `DistributeCPUsAcrossNUMA` I think.

ffromani commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -138,6 +148,10 @@ func NewStaticPolicyOptions(policyOptions map[string]string) (StaticPolicyOption
 		return opts, fmt.Errorf("static policy options %s and %s can not be used at the same time", DistributeCPUsAcrossNUMAOption, DistributeCPUsAcrossCoresOption)
 	}
 
+	if opts.AlignByUnCoreCacheOption && opts.DistributeCPUsAcrossCores {
```
IIUC we are always compatible with `AlignBySocket` considering how available hardware (x86, arm) is engineered, correct?

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -495,6 +495,12 @@ func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs int)
 		return takeByTopologyNUMADistributed(p.topology, availableCPUs, numCPUs, cpuGroupSize, cpuSortingStrategy)
 	}
+
+	if p.options.AlignByUnCoreCacheOption {
+
+		return takeByTopologyUnCoreCachePacked(p.topology, availableCPUs, numCPUs, cpuSortingStrategy)
+
+	}
```
we need to merge this logic in  `takeByTopologyNUMAPacked` passing down the flag where needed, need to think about it

ajcaldelas replied:
Last week we were taking a look at this and we don't think this can be merged because this function is called specifically based on the policy option, we would need to be able to check inside this function right?

ffromani replied:
yes, we can. I see this was fixed in the last upload.

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology.go:
```diff
@@ -108,6 +113,73 @@ func (d CPUDetails) KeepOnly(cpus cpuset.CPUSet) CPUDetails {
+	}
+	sort.Ints(coreIDs)
+
+	// return only unique coreID
+	var coresNeeded []int
+
+	if len(coreIDs) > numCoresNeeded {
+		// return only what is needed
+		coresNeeded = coreIDs[0:numCoresNeeded]
+	} else {
+		// return the full list
+		coresNeeded = coreIDs
+	}
+	klog.V(2).InfoS("Available coreIDs : ", "coresNeeded", coresNeeded)
+	return cpuset.New(coresNeeded...)
```
why do we need this logic?

wongchar replied:
Before this portion of code, the function allocates fully available UncoreCaches to a container if the needed CPUs are greater or equal to the size of a fully available UncoreCache.

The highlighted logic handles the case where number of CPUs needed (or remainder) is less than a fully available UncoreCache.

This portion of the logic is important is for the following:

- Minimizing the distribution across number of UncoreCaches will provide better performance. There are cases where skipping this logic and going to `takeFullCores` could result in distribution across more UncoreCaches.
- On smaller core count processors, there are even fewer UncoreCaches to align to. (Ex: 32 cores and only 4 UncoreCaches). Smaller processors commonly deploy smaller containers with CPU requirements less than an UncoreCache and would not benefit (ex: edge/telco use cases).

This logic currently matches the corresponding KEP. I understand that this may deviate from the logic flow of takeFirstLevel and takeSecondLevel, but I think without it, there would be performance optimizations left on the table.

Happy to discuss further, please let me know your thoughts

ffromani replied:
OK, makes sense but the code it's a bit too hard to follow. Is the following simplification correct?
```
diff --git a/pkg/kubelet/cm/cpumanager/topology/topology.go b/pkg/kubelet/cm/cpumanager/topology/topology.go
index 777fffb5483..f33a41a5546 100644
--- a/pkg/kubelet/cm/cpumanager/topology/topology.go
+++ b/pkg/kubelet/cm/cpumanager/topology/topology.go
@@ -18,8 +18,6 @@ package topology
 
 import (
        "fmt"
-       "slices"
-       "sort"
 
        cadvisorapi "github.com/google/cadvisor/info/v1"
        "k8s.io/klog/v2"
@@ -137,33 +135,26 @@ func (d CPUDetails) UncoreInNUMANodes(ids ...int) cpuset.CPUSet {
 }
 
 // CoresNeededInUncoreCache returns either the full list of all available unique core IDs associated with the given
-// UnCoreCache IDs in this CPUDetails or subset that matches the ask.
+// UncoreCache IDs in this CPUDetails or subset that matches the ask.
 func (d CPUDetails) CoresNeededInUncoreCache(numCoresNeeded int, ids ...int) cpuset.CPUSet {
+       coreIDs := d.coresInUncoreCache(ids...)
+       if coreIDs.Size() <= numCoresNeeded {
+               return coreIDs
+       }
+       tmpCoreIDs := coreIDs.List()
+       return cpuset.New(tmpCoreIDs[:numCoresNeeded]...)
+}
+
+func (d CPUDetails) coresInUncoreCache(ids ...int) cpuset.CPUSet {
        var coreIDs []int
        for _, id := range ids {
                for _, info := range d {
                        if info.UncoreCacheID == id {
-
-                               if !slices.Contains(coreIDs, info.CoreID) {
-                                       coreIDs = append(coreIDs, info.CoreID)
-                               }
+                               coreIDs = append(coreIDs, info.CoreID)
                        }
                }
        }
-       sort.Ints(coreIDs)
-
-       // return only unique coreID
-       var coresNeeded []int
-
-       if len(coreIDs) > numCoresNeeded {
-               // return only what is needed
-               coresNeeded = coreIDs[0:numCoresNeeded]
-       } else {
-               // return the full list
-               coresNeeded = coreIDs
-       }
-       klog.V(2).InfoS("Available coreIDs : ", "coresNeeded", coresNeeded)
-       return cpuset.New(coresNeeded...)
+       return cpuset.New(coreIDs...)
 }
 
 // CPUsInUncoreCaches returns all the logical CPU IDs associated with the given
```
(note: AFAIK it passes current unit tests)

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology.go:
```diff
@@ -260,9 +332,10 @@ func Discover(machineInfo *cadvisorapi.MachineInfo) (*CPUTopology, error) {
+						CoreID:        coreID,
+						SocketID:      core.SocketID,
+						NUMANodeID:    node.Id,
+						UnCoreCacheID: core.UncoreCaches[0].Id,
```
this can break in some server-grade CPUs which don't expose uncore cache. I suggest to replace with something like
```
const (
	NullUncoreCacheID int = 65535 // TODO: "high enough" value to signal "impossible"
)

func getUncoreCacheID(uncoreCaches []cadvisorapi.Cache) int {
	if len(uncoreCaches) < 1 {
		return NullUncoreCacheID
	}
	// Even though cadvisor API returns a slice, we only expect either 0 or 1 uncore caches,
	// so everything past first entry should be discarded
	return uncoreCaches[0].Id
}

// ...

UnCoreCacheID: getUncoreCacheID(core.UncoreCaches),
```

we may want to use a better ID to signal "no uncore caches detected

ajcaldelas replied:
I'm not totally sure why, but can't we presume it will be `0`? I think that's the value found in `/sys/device/system/cpu/cpuX/cache/index3/id`

ffromani replied:
seen myself that in some cases (= with some commonly available xeons) `cadvisor` just gives us an empty slice, which will result in nil pointer deference.

I usually prefer fixing the bug in the right layer rather than add defensive programming, but engineering it's all about tradeoffs, so in this case I'd start adding in our code some helper function like I described, and then (in parallel/later) go review and fix `cadvisor`.

ajcaldelas replied:
I agree, for now I added the function but if we see an empty slice I return `0`

ajcaldelas replied:
After reviewing this should fallback to the socketid or to the node.caches[0] value. For now we left it at the socketID.

wongchar replied:
Added a test case to handle empty UncoreCacheIDs in the Discover function

ffromani replied:
using SocketID as fallback SGTM, thanks!

ffromani reviewed (commented):
initial review

kannon92 commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -502,6 +502,12 @@ func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs int)
 		return takeByTopologyNUMADistributed(p.topology, availableCPUs, numCPUs, cpuGroupSize, cpuSortingStrategy)
 	}
+
+	if p.options.PreferAlignByUnCoreCacheOption {
```
Could you add some unit tests around this?

I want to make sure that the fallback (if feature gate is turned off) is covered with some unit tests.

ffromani replied:
Tests are surely welcome, but I think we will need to pass this flag somehow deep down the takeByTopologyNUMAPacked, something like we did for `cpuSortinStrategy` in previous releases.

In any case, I don't think we should proliferate further the `takeByTopology` functions.

kannon92 replied:
We have a file for adding tests for policy_static_test. Why can't we at least add some tests based on the right options. 

We should have testing around the case if this feature gate is turned off. We probably won't cover this for e2e tests but we should at least make sure that this feature enablement/disenablement is covered.

ffromani replied:
I fully agree.

ffromani replied:
this seems sufficient for non-regression testing, yes.

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology.go:
```diff
@@ -317,6 +317,15 @@ func (d CPUDetails) CPUsInCores(ids ...int) cpuset.CPUSet {
 
+func getUncoreCacheID(uncoreCaches []cadvisorapi.Cache) int {
+	if len(uncoreCaches) < 1 {
+		return 0
```
please add a comment to explain why we pick 0. One-liner is sufficient.

In addition, we can use a couple unit tests in `topology_test.go`. Either feed a topology object wit nil/empty slice (harder, but preferred), or add target tests for this helper function (simpler, but uglier)

ajcaldelas replied:
changed this, and added unit testing, will use socket id for the uncorecache

wongchar replied:
Added a test to handle empty UncoreCache slice in Discover and test for the getUncoreCacheID function

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +575,50 @@ func (a *cpuAccumulator) takeFullSockets() {
+		// determine if SMT is enabled
+		numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore()
+
+		var freeCPUsInUncoreCache cpuset.CPUSet
```
ok, it took me a while to eventually realize this code implements what we agreed on the KEP, which in turn made me realize that we can use a simplification of the algorithm in the KEP, or at very least have a review of it and explain why we need to diverge from the flow here.
Since we are gathering quite a lot of todos, we may want to have another alpha cycle in 1.33. Let's see.

wongchar replied:
Responded to first part in previous post.
No opposition to remain in alpha in 1.33 to iron more out.

ffromani reviewed (commented):
another round of review, quite some polishing needed.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -184,6 +207,18 @@ func (s *socketsFirst) sortAvailableSockets() []int {
+		result = append(result, uncore...)
+	}
+	return result
+}
```
uhm, makes sense. Will look again more carefully.

ffromani replied:
So, in both cases `sortAvailableUncoreCaches` uses the same logic but depends on the behavior of `sortAvailableNUMANodes`, and only the latter will change its behavior depending if `numaFirst` or `socketFirst`, correct?
If so, this means that `sortAvailableUncoreCaches` per se should not be in the `numaOrSocketsFirstFuncs` interface, because the logic doesn't change; only the logic of its dependencies do.
This also seems what an admittedly quick re-read of the KEP is confirming (will look carefully again later).

If all the above holds, we can simplify the code like this:
```
diff --git a/pkg/kubelet/cm/cpumanager/cpu_assignment.go b/pkg/kubelet/cm/cpumanager/cpu_assignment.go
index f99d92f7aac..597bb9fd5f3 100644
--- a/pkg/kubelet/cm/cpumanager/cpu_assignment.go
+++ b/pkg/kubelet/cm/cpumanager/cpu_assignment.go
@@ -97,7 +97,6 @@ type numaOrSocketsFirstFuncs interface {
 	takeFullSecondLevel()
 	sortAvailableNUMANodes() []int
 	sortAvailableSockets() []int
-	sortAvailableUncoreCaches() []int
 	sortAvailableCores() []int
 }
 
@@ -119,18 +118,6 @@ func (n *numaFirst) takeFullSecondLevel() {
 	n.acc.takeFullSockets()
 }
 
-// If NUMA nodes are higher in the memory hierarchy than sockets, then just
-// sort the UncoreCache within the socket and return them.
-func (n *numaFirst) sortAvailableUncoreCaches() []int {
-	var result []int
-	for _, socket := range n.acc.sortAvailableNUMANodes() {
-		uncore := n.acc.details.UncoreInNUMANodes(socket).UnsortedList()
-		n.acc.sort(uncore, n.acc.details.CPUsInUncoreCaches)
-		result = append(result, uncore...)
-	}
-	return result
-}
-
 // If NUMA nodes are higher in the memory hierarchy than sockets, then just
 // sort the NUMA nodes directly, and return them.
 func (n *numaFirst) sortAvailableNUMANodes() []int {
@@ -197,18 +184,6 @@ func (s *socketsFirst) sortAvailableSockets() []int {
 	return sockets
 }
 
-// If sockets higher in the memory hierarchy than NUMA nodes, then UncoreCache
-// sit directly below NUMA Nodes in the memory hierchy
-func (s *socketsFirst) sortAvailableUncoreCaches() []int {
-	var result []int
-	for _, uncore := range s.acc.sortAvailableNUMANodes() {
-		uncore := s.acc.details.UncoreInNUMANodes(uncore).UnsortedList()
-		s.acc.sort(uncore, s.acc.details.CPUsInUncoreCaches)
-		result = append(result, uncore...)
-	}
-	return result
-}
-
 // If sockets are higher in the memory hierarchy than NUMA nodes, then cores
 // sit directly below NUMA Nodes in the memory hierarchy.
 func (s *socketsFirst) sortAvailableCores() []int {
@@ -377,10 +352,20 @@ func (a *cpuAccumulator) freeSockets() []int {
 	return free
 }
 
+func (a *cpuAccumulator) sortAvailableUncoreCaches() []int {
+	var result []int
+	for _, numa := range a.sortAvailableNUMANodes() {
+		uncore := a.details.UncoreInNUMANodes(numa).UnsortedList()
+		a.sort(uncore, a.details.CPUsInUncoreCaches)
+		result = append(result, uncore...)
+	}
+	return result
+}
+
 // Returns free UncoreCache IDs as a slice sorted by sortAvailableUnCoreCache().
 func (a *cpuAccumulator) freeUncoreCache() []int {
 	free := []int{}
-	for _, uncore := range a.numaOrSocketsFirst.sortAvailableUncoreCaches() {
+	for _, uncore := range a.sortAvailableUncoreCaches() {
 		if a.isUncoreCacheFree(uncore) {
 			free = append(free, uncore)
 		}
@@ -573,7 +558,7 @@ func (a *cpuAccumulator) takeFullUnCore() {
 // First try to take full UncoreCache, if available and need is at least the size of the UncoreCache group.
 // Second try to take the partial UncoreCache if available and the request size can fit w/in the UncoreCache.
 func (a *cpuAccumulator) takeUncoreCache() {
-	for _, uncore := range a.numaOrSocketsFirst.sortAvailableUncoreCaches() {
+	for _, uncore := range a.sortAvailableUncoreCaches() {
 		// take full UncoreCache if the CPUs needed is greater than free UncoreCache size
 		if a.numCPUsNeeded >= a.topo.NumCPUs/a.topo.NumUncoreCache {
 			a.takeFullUnCore()
```

note thispatch  assumes both my other simplifications proposals have been applied, so we don't have `takeThirdLevel` anymore.

wongchar replied:
This makes sense, applied these changes to have one `sortAvailableUncoreCaches` by NUMA.
Removed takeThirdLevel as well to simplify.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -346,6 +387,22 @@ func (a *cpuAccumulator) freeSockets() []int {
+}
+
+// Returns all UncoreCache IDs as a slice sorted by sortAvailableUncoreCache().
+func (a *cpuAccumulator) allUncoreCache() []int {
+	return a.numaOrSocketsFirst.sortAvailableUncoreCaches()
+}
```
this is called once, and it's one-liner. Let's inline it on its one and only call site.

wongchar replied:
Removed allUncoreCache and called the single UncoreCache sorting function directly.

ajcaldelas commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +575,50 @@ func (a *cpuAccumulator) takeFullSockets() {
+		for _, coreID := range freeCoresInUncoreCache.List() {
+			freeCPUsInUncoreCache = freeCPUsInUncoreCache.Union(a.topo.CPUDetails.CPUsInCores(coreID))
+		}
+		klog.V(2).InfoS("freeCPUsInUncorecache  : ", "freeCPUsInUncorecache", freeCPUsInUncoreCache)
```
done

ajcaldelas commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +575,50 @@ func (a *cpuAccumulator) takeFullSockets() {
+		var freeCPUsInUncoreCache cpuset.CPUSet
+		// when CPUs needed is smaller than a free UncoreCache size, try to assign cores from a single UncoreCache
+		freeCoresInUncoreCache := a.details.CoresNeededInUncoreCache(numCoresNeeded, uncore)
+		klog.V(2).InfoS("free cores from a.details list: ", "freeCoresInUncorecache", freeCoresInUncoreCache)
```
done

wongchar commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -118,6 +120,23 @@ func (n *numaFirst) takeFullSecondLevel() {
 	n.acc.takeFullSockets()
 }
 
+// In Split UncoreCache Topology, we take from the sets of UncoreCache as the third level
```
takeThirdLevel removed per your recommendation :)

wongchar commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -318,6 +353,12 @@ func (a *cpuAccumulator) isSocketFree(socketID int) bool {
 }
 
+// Returns true if the supplied UnCoreCache is fully available in `a.details`.
+// "fully available" means that all the CPUs in it are free.
```
Changes applied

wongchar commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -502,7 +502,8 @@ func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs int)
 	}
-	return takeByTopologyNUMAPacked(p.topology, availableCPUs, numCPUs, cpuSortingStrategy)
+
+	return takeByTopologyNUMAPacked(p.topology, availableCPUs, numCPUs, cpuSortingStrategy, p.options.PreferAlignByUncoreCacheOption)
```
Agreed, no opposition to remain in alpha for the next cycle as well to address (v1.33)

wongchar commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +575,50 @@ func (a *cpuAccumulator) takeFullSockets() {
+			return
+		}
+
+		// determine if SMT is enabled
```
Clarified the comment, please let me know your thoughts

wongchar commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -118,6 +120,23 @@ func (n *numaFirst) takeFullSecondLevel() {
+// In Split UncoreCache Topology, we take from the sets of UncoreCache as the third level
+func (n *numaFirst) takeThirdLevel() {
+	n.acc.takeUncoreCache()
+}
```
Makes sense. Removed takeThirdLevel.

ffromani commented:
/retest

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +546,51 @@ func (a *cpuAccumulator) takeFullSockets() {
+func (a *cpuAccumulator) takeUncoreCache() {
+	for _, uncore := range a.sortAvailableUncoreCaches() {
+		// take full UncoreCache if the CPUs needed is greater than free UncoreCache size
+		if a.numCPUsNeeded >= a.topo.NumCPUs/a.topo.NumUncoreCache {
```
please replace this `if` with
```
if a.needsAtLeast(a.topo.NumCPUs/a.topo.NumUncoreCache)
```
in addition: `a.topo.NumCPUs/a.topo.NumUncoreCache` is a loop invariant, should be computed once and at the top of the function and then the value reused.
Last but not least: this is the wrong way to count the (V)CPUs per Uncore group. It assumes equal uncore groups, but we know already this is overly simplified (offlined cpus is a easy example, and yes, we got actual reports about this) but still it's sadly the best we can do in the current architecture.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +546,51 @@ func (a *cpuAccumulator) takeFullSockets() {
+
+		// if SMT is enabled, number of cores needed is half the number of CPUs needed
+		// if SMT is disabled, number of cores needed is equal to number of CPUs
+		numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore()
```
part of the confusion is we keep mixing physical cpus (virtual cpu pairs) with virtual cpus. We should be uniform in this function.

ajcaldelas replied:
AFAIK  I thought that when they mention CPUs to assume it is always a vCPU or logical CPU.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +546,51 @@ func (a *cpuAccumulator) takeFullSockets() {
+
+		var freeCPUsInUncoreCache cpuset.CPUSet
+		// when CPUs needed is smaller than a free UncoreCache size, try to assign cores from a single UncoreCache
+		freeCoresInUncoreCache := a.details.CoresNeededInUncoreCache(numCoresNeeded, uncore)
```
ok, following the kube terminlogy we return cores aka virtual cpus

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +546,51 @@ func (a *cpuAccumulator) takeFullSockets() {
+		freeCoresInUncoreCache := a.details.CoresNeededInUncoreCache(numCoresNeeded, uncore)
+		klog.V(4).InfoS("free cores from a.details list: ", "freeCoresInUncorecache", freeCoresInUncoreCache)
+		for _, coreID := range freeCoresInUncoreCache.List() {
+			freeCPUsInUncoreCache = freeCPUsInUncoreCache.Union(a.topo.CPUDetails.CPUsInCores(coreID))
```
and here we go back to cpus aka pohysical cpus. We should simplify this code and always reason in terms of cores or cpus. Let's review what other functions do and act accordingly.

ajcaldelas replied:
We could potentially split that out into another helper function

ffromani reviewed (commented):
we are indeed converging, but we need some polishing and simplification

ffromani commented:
EDIT: mixed up myself core vs cpu (physical vs virtual). We need to simplify this code.

ffromani commented:
thanks for addressing my comments, glad I'm getting the flow right.

So the code is doing what the KEP declared, and we are converging towards another iteration in alpha in 1.33 to streamline more and review based on the 1.32 feedback.

After the last changes, considered that `takeUncoreCache` does *not* depend on the numaFist/socketFirst memory hierarchy (and this took me a while to realize), the last concern is about `takeUncoreCache` implementation. 

Once these concerns are figured out, and once the unit test coverage is reviewed and good enough (we should not be too far from this) this change LGTM and we can move on. 

So I took quite some time to review carefully the code and I came up with a series of changes to simplify the code. Unit tests keep passing after each step, so I should be on a correct track. I'm posting the series here to both suggest simplifications and to confirm my understanding.

If you folks take these suggestion, feel free to squash them in a single change. We should keep the number of commits minimal, and there's no point in keep the detailed steps in the final changeset.

ffromani commented:
first, consolidtating the logs, let's minimize the call to `klog`. This is trivial, posting for the sake of making the next changes in the series consistent:
```
diff --git a/pkg/kubelet/cm/cpumanager/cpu_assignment.go b/pkg/kubelet/cm/cpumanager/cpu_assignment.go
index 7cb1d0e9109..9d02b7269c8 100644
--- a/pkg/kubelet/cm/cpumanager/cpu_assignment.go
+++ b/pkg/kubelet/cm/cpumanager/cpu_assignment.go
@@ -576,16 +576,13 @@ func (a *cpuAccumulator) takeUncoreCache() {
 		var freeCPUsInUncoreCache cpuset.CPUSet
 		// when CPUs needed is smaller than a free UncoreCache size, try to assign cores from a single UncoreCache
 		freeCoresInUncoreCache := a.details.CoresNeededInUncoreCache(numCoresNeeded, uncore)
-		klog.V(4).InfoS("free cores from a.details list: ", "freeCoresInUncorecache", freeCoresInUncoreCache)
 		for _, coreID := range freeCoresInUncoreCache.List() {
 			freeCPUsInUncoreCache = freeCPUsInUncoreCache.Union(a.topo.CPUDetails.CPUsInCores(coreID))
 		}
-		klog.V(4).InfoS("freeCPUsInUncorecache  : ", "freeCPUsInUncorecache", freeCPUsInUncoreCache)
+		klog.V(4).InfoS("takeUncoreCache", "freeCoresInUncoreCache", freeCoresInUncoreCache.String(), "freeCPUsInUncorecache", freeCPUsInUncoreCache.String())
 		if a.numCPUsNeeded == freeCPUsInUncoreCache.Size() {
-			klog.V(4).InfoS("takePartialUncore: claiming cores from Uncorecache ID", "uncore", uncore)
 			a.take(freeCPUsInUncoreCache)
 		}
-
 		if a.isSatisfied() {
 			return
 		}
-- 
2.46.2
```

Then, no need to sort explicitely, we can (and should) reuse what we have already in the `cpuset` package, which we heavily depend on anyway. This helps to make the code more idiomatic:
```
diff --git a/pkg/kubelet/cm/cpumanager/topology/topology.go b/pkg/kubelet/cm/cpumanager/topology/topology.go
index 405771ce5cc..b19a272a6e1 100644
--- a/pkg/kubelet/cm/cpumanager/topology/topology.go
+++ b/pkg/kubelet/cm/cpumanager/topology/topology.go
@@ -18,8 +18,6 @@ package topology
 
 import (
 	"fmt"
-	"slices"
-	"sort"
 
 	cadvisorapi "github.com/google/cadvisor/info/v1"
 	"k8s.io/klog/v2"
@@ -158,14 +156,11 @@ func (d CPUDetails) coresInUncoreCache(ids ...int) []int {
 	for _, id := range ids {
 		for _, info := range d {
 			if info.UncoreCacheID == id {
-				if !slices.Contains(coreIDs, info.CoreID) {
-					coreIDs = append(coreIDs, info.CoreID)
-				}
+				coreIDs = append(coreIDs, info.CoreID)
 			}
 		}
 	}
-	sort.Ints(coreIDs)
-	return coreIDs
+	return cpuset.New(coreIDs...).List()
 }
 
 // CPUsInUncoreCaches returns all the logical CPU IDs associated with the given
-- 
2.46.2
```

Now: List() implies the sorting. I'd do the relatively expensive sort operation only when really needed, and extract a subset of cpus from the total availability only if needed (not all the time)
```
diff --git a/pkg/kubelet/cm/cpumanager/topology/topology.go b/pkg/kubelet/cm/cpumanager/topology/topology.go
index b19a272a6e1..d39d0ff4298 100644
--- a/pkg/kubelet/cm/cpumanager/topology/topology.go
+++ b/pkg/kubelet/cm/cpumanager/topology/topology.go
@@ -138,20 +138,14 @@ func (d CPUDetails) UncoreInNUMANodes(ids ...int) cpuset.CPUSet {
 // UnCoreCache IDs in this CPUDetails or subset that matches the ask.
 func (d CPUDetails) CoresNeededInUncoreCache(numCoresNeeded int, ids ...int) cpuset.CPUSet {
 	coreIDs := d.coresInUncoreCache(ids...)
-	// return only unique coreID
-	var coresNeeded []int
-	if len(coreIDs) > numCoresNeeded {
-		// return only what is needed
-		coresNeeded = coreIDs[0:numCoresNeeded]
-	} else {
-		// return the full list
-		coresNeeded = coreIDs
+	if coreIDs.Size() > numCoresNeeded {
+		tmpCoreIDs := coreIDs.List()
+		return cpuset.New(tmpCoreIDs[:numCoresNeeded]...)
 	}
-	klog.V(2).InfoS("Available coreIDs", "coresNeeded", coresNeeded)
-	return cpuset.New(coresNeeded...)
+	return coreIDs
 }
 
-func (d CPUDetails) coresInUncoreCache(ids ...int) []int {
+func (d CPUDetails) coresInUncoreCache(ids ...int) cpuset.CPUSet {
 	var coreIDs []int
 	for _, id := range ids {
 		for _, info := range d {
@@ -160,7 +154,7 @@ func (d CPUDetails) coresInUncoreCache(ids ...int) []int {
 			}
 		}
 	}
-	return cpuset.New(coreIDs...).List()
+	return cpuset.New(coreIDs...)
 }
 
 // CPUsInUncoreCaches returns all the logical CPU IDs associated with the given
-- 
2.46.2
```

(to be continued...)

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -318,6 +329,12 @@ func (a *cpuAccumulator) isSocketFree(socketID int) bool {
 	return a.details.CPUsInSockets(socketID).Size() == a.topo.CPUsPerSocket()
 }
 
+// Returns true if the supplied UnCoreCache is fully available,
+// meaning all its CPUs are free in `a.details`.
```
ook, my fault, I didn't realize we were following the example of similar function in the file. Please let's go abck to
```
// Returns true if the supplied uncore cache group is fully available in `a.details`.
// "fully available" means that all the CPUs in it are free.
```

ajcaldelas replied:
done

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -518,6 +546,59 @@ func (a *cpuAccumulator) takeFullSockets() {
 		a.take(cpusInSocket)
 	}
 }
+func (a *cpuAccumulator) takeFullUncore() {
```
please add a space to separate the functions.

ajcaldelas replied:
done

ffromani reviewed (commented):
looks good, thanks for implementing the suggestion. Please squash as much as you can the commits.

I think we can now doublecheck the test coverage is sufficient, if so I think we are good enough for the first iteration of alpha stage in 1.32

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology.go:
```diff
@@ -108,6 +120,67 @@ func (d CPUDetails) KeepOnly(cpus cpuset.CPUSet) CPUDetails {
+}
+
+// Helper function that just gets the cores
+func (d CPUDetails) CoresInUncoreCache(ids ...int) cpuset.CPUSet {
```
nit: this can be private

ajcaldelas replied:
Changed to private, did not create a test for this function either

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology.go:
```diff
@@ -245,6 +318,16 @@ func (d CPUDetails) CPUsInCores(ids ...int) cpuset.CPUSet {
 
+func getUncoreCacheID(core cadvisorapi.Core) int {
+	if len(core.UncoreCaches) < 1 {
+		// In case cAdvisor is nil, failback to socket alignment since uncorecache is not shared
+		return core.SocketID
```
unless there are clearly better options I'd stay like this for 1.32

ffromani reviewed (commented):
codewise LGTM, need to review the unit test coverage and we can move forward

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment_test.go:
```diff
@@ -668,6 +668,79 @@ func TestTakeByTopologyNUMAPacked(t *testing.T) {
+			"",
+			cpuset.New(1, 2, 4, 5, 6, 7, 8, 9, 10, 11),
+		},
+		{
+			"take one cpu from dual socket with HT - core from Socket 0",
+			topoDualSocketHT,
+			StaticPolicyOptions{PreferAlignByUncoreCacheOption: true},
+			cpuset.New(1, 2, 3, 4, 5, 7, 8, 9, 10, 11),
+			1,
+			"",
+			cpuset.New(2),
+		},
```
what's the purpose of this specific case? the used topology doesn't seem to have split L3 cache. Let's clarify and fix (not drop if possible, only if redundant)

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment_test.go:
```diff
@@ -668,6 +668,79 @@ func TestTakeByTopologyNUMAPacked(t *testing.T) {
+			4,
+			"",
+			cpuset.New(8, 9, 10, 11),
+		},
```
I'd add another case to check what happens when we have partial allocations, not only full allocations.
If this overlaps with another test (which one?) let's just add a comment instead.
```
+               {
+                       "take first available UncoreCache from second socket, partially allocated UncoreCache groups in first socket",
+                       topoUncoreDualSocketNoSMT,
+                       StaticPolicyOptions{PreferAlignByUncoreCacheOption: true},
+                       mustParseCPUSet(t, "1-3,5-15"),
+                       4,
+                       "",
+                       cpuset.New(8, 9, 10, 11),
+               },
```

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment_test.go:
```diff
@@ -668,6 +668,79 @@ func TestTakeByTopologyNUMAPacked(t *testing.T) {
+			"take cpus from best available UncoreCache group of multi uncore cache single socket - SMT enabled",
+			topoUncoreSingleSocketSMT,
+			StaticPolicyOptions{PreferAlignByUncoreCacheOption: true},
+			mustParseCPUSet(t, "2-3,10-11,4-7,12-15"),
```
this is unnecessarily hard to read, any reason why is not `mustParseCPUSet(t, "2-7,10-15")` ?

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment_test.go:
```diff
@@ -668,6 +668,79 @@ func TestTakeByTopologyNUMAPacked(t *testing.T) {
+			mustParseCPUSet(t, "1-7,9-15"),
+			10,
+			"",
+			mustParseCPUSet(t, "4-7,12-15,1,9"),
```
please add a comment to document that this is the expected allocation order.
This is fully equivalent to `mustParseCPUSet(t, "1,4-7,9,12-15")` but I like the reminder of allocation ordering

ffromani reviewed (commented):
LGTM once the comments about tests are addressed.

I'd like more unit tests, let's keep iterating about that, but if nothing obvious comes to mind we can add them in followup PRs.

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology_test.go:
```diff
@@ -33,6 +33,41 @@ func Test_Discover(t *testing.T) {
 		want        *CPUTopology
 		wantErr     bool
 	}{
+		{
+			name: "EmptyUncoreCache",
+			machineInfo: cadvisorapi.MachineInfo{
+				NumCores:   8,
+				NumSockets: 1,
+				Topology: []cadvisorapi.Node{
+					{Id: 0,
+						Cores: []cadvisorapi.Core{
+							{SocketID: 0, Id: 0, Threads: []int{0, 4}, UncoreCaches: []cadvisorapi.Cache{}},
+							{SocketID: 0, Id: 1, Threads: []int{1, 5}, UncoreCaches: []cadvisorapi.Cache{}},
+							{SocketID: 0, Id: 2, Threads: []int{2, 6}, UncoreCaches: []cadvisorapi.Cache{}},
+							{SocketID: 0, Id: 3, Threads: []int{3, 7}, UncoreCaches: []cadvisorapi.Cache{}},
+						},
+					},
+				},
+			},
+			want: &CPUTopology{
+				NumCPUs:        8,
+				NumSockets:     1,
+				NumCores:       4,
+				NumNUMANodes:   1,
+				NumUncoreCache: 1,
+				CPUDetails: map[int]CPUInfo{
+					0: {CoreID: 0, SocketID: 0, NUMANodeID: 0, UncoreCacheID: 0},
+					1: {CoreID: 1, SocketID: 0, NUMANodeID: 0, UncoreCacheID: 0},
+					2: {CoreID: 2, SocketID: 0, NUMANodeID: 0, UncoreCacheID: 0},
+					3: {CoreID: 3, SocketID: 0, NUMANodeID: 0, UncoreCacheID: 0},
+					4: {CoreID: 0, SocketID: 0, NUMANodeID: 0, UncoreCacheID: 0},
+					5: {CoreID: 1, SocketID: 0, NUMANodeID: 0, UncoreCacheID: 0},
```
we also need a test for `nil` uncore cache, when the `UncoreCaches` slice is `nil`

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology_test.go:
```diff
@@ -1100,3 +1414,38 @@ func TestCPUNUMANodeID(t *testing.T) {
+			},
+			want: 5,
+		},
+		{
+			name: "Core with empty uncore cache info",
+			args: cadvisorapi.Core{
+				SocketID:     2,
+				UncoreCaches: []cadvisorapi.Cache{},
+			},
+			want: 2,
+		},
```
please add a testcase when `UncoreCaches` is `nil`

ajcaldelas replied:
done

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology_test.go:
```diff
@@ -1100,3 +1414,38 @@ func TestCPUNUMANodeID(t *testing.T) {
 	}
 }
+
+func Test_getUncoreCacheID(t *testing.T) {
```
I'd like to see tests also for the quite a few helper functions we added, like (in no particular order nor claim of completeness):
- `CPUsPerUncore`
- `UncoreCaches`
- `UncoreInNUMANodes`
- `CPUsInUncoreCaches`
- `CoresNeededInUncoreCache` (not needed if we make it private)

no need to write like 100 test cases, but considering we have a quite rich collection of topologies, we can add few tests for them

ajcaldelas replied:
Quick question then, `getUncoreCacheID` is a private function should I remove this test?

ajcaldelas replied:
Added Unit tests 
- `CPUsPerUncore`
- `UncoreCaches`
- `UncoreInNUMANodes`
- `CPUsInUncoreCaches`

ffromani replied:
> Quick question then, `getUncoreCacheID` is a private function should I remove this test?

let's keep it for now. We can make exceptions, but these should be, well, exceptional :)

ffromani reviewed (commented):
LGTM, thanks for adding tests
good enough start. Please squash commits and IMO we are done.

ffromani commented:
/lgtm

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/assign @klueska

ffromani commented:
/release-note introduce a new CPU Manager static policy option that makes best effort to align CPU resources by L3 (last-level) cache.

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology_test.go:
```diff
@@ -33,6 +33,41 @@ func Test_Discover(t *testing.T) {
 		want        *CPUTopology
 		wantErr     bool
 	}{
+		{
```
just noticed the file permissions got changed. Please revert the file permission to 0644 (remove the execution bits) if you need to resubmit

ajcaldelas replied:
Sorry this was missed will change this now

ffromani reviewed (commented):
/lgtm

rebase seems OK now

ffromani commented on pkg/kubelet/cm/cpumanager/topology/topology.go:
```diff
@@ -90,11 +101,12 @@ func (topo *CPUTopology) CPUNUMANodeID(cpu int) (int, error) {
 }
 
-// CPUInfo contains the NUMA, socket, and core IDs associated with a CPU.
+// CPUInfo contains the NUMA, socket, unCoreCache and core IDs associated with a CPU.
```
nit: `UncoreCache`

ffromani reviewed (commented):
/lgtm

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
gentle ping because the code freeze is looming

@mrunalp / @klueska / @SergeyKanzhelev

mrunalp commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -95,10 +95,8 @@ func standardDeviation(xs []int) float64 {
 type numaOrSocketsFirstFuncs interface {
 	takeFullFirstLevel()
 	takeFullSecondLevel()
-	takeThirdLevel()
```
This is added in the first commit and removed in the second. Can we make the commits additive or squash them?

ffromani replied:
That would be great indeed, maybe the only concern is properly preserving authorship since there are 3 contributors to this PR? Maybe I'm overthinking (can't recall the fine details of the CLA). If this is not a concern and/or can be solved by you folks contributing the change, by all means please squash/rearrange the commits to make them additive!

ajcaldelas replied:
I’ll check with our group but I believe you are correct that this will remove an author, which may not be desired unfortunately

MarSik commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -519,6 +547,60 @@ func (a *cpuAccumulator) takeFullSockets() {
+	}
+}
+
+func (a *cpuAccumulator) takePartialUncore(uncoreID int) {
```
The way I read this method it should be called `takeRestOfUncore` right?

It seems to be an optimization that tries to prevent leaving unusably small amount of cpus lying around free. Am I right?

ajcaldelas replied:
@wongchar correct me if I'm wrong, but that's not quite what's it's trying to do, it will check to see if we have some available to the uncore cache group and take what it needs but not exactly taking the rest of the uncore.

MarSik replied:
Well.. maybe I do not understand the logic properly, but the if on row 569 seems to be checking whether the needed cpu count == (equals) the free cpus count.

Granted, the freeCores/freeCPUs computation is based on numCoresNeeded, but the semantics of the whole sequence is pretty complicated.

Maybe a better comment is needed.

MarSik replied:
I did a few tests and the logic of the PR seems to be able to claim partial uncores overall, it is the naming and logic is not obvious from the code.

ffromani replied:
The code (as I understood and LGTM'd) is meant to allocat a partially-available uncore, with some cpus already busy, depending on both the free cpus on the uncore and the requested quantity. The name should IMO contain somehow "partial" or a synonim, because after this code runs, it legitimately leave a uncore group partially allocated.

wongchar replied:
Agree with comments above.

`takePartialUncore` is used when the number of cpus needed is less than a free uncore cache size (or when the remainder of cpus needed after `takeFullUncore` is less than a free uncore cache size). This is a best-effort method to keep the cpu assignment limited to a single uncore cache. 

Passing directly to `takeFullCores` or `takeRemainingCPUs` immediately after `takeFullUncore` could result in cpu assignments distributing across multiple uncore caches when it could fit within a single uncore cache.

Added some comments to `takePartialUncore` and 'takeUncoreCache' to help outline the comments above. Please let me know if this does not clear it up :)

MarSik commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -118,6 +118,17 @@ func (n *numaFirst) takeFullSecondLevel() {
 	n.acc.takeFullSockets()
 }
 
+// Sort the UncoreCaches within the NUMA nodes.
```
Please improve the comment to make the sorting intent clear.

if I am not mistaken, this sorts in ascending order based on the free cpus in each uncore cache group? This is related to the takeUncoreCache algorithm.

MarSik replied:
Ok, I was mistaken, it sorts based on CPU IDs which matches the KEP. Sorry for the noise.

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

mrunalp commented:
/hold
for @MarSik comments

wongchar committed:
update comments

ffromani commented:
/test pull-kubernetes-e2e-gce

ffromani commented:
/lgtm

ffromani commented:
/hold cancel

comments have been addressed

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

MarSik commented:
/lgtm thanks for explaining the more complicated pieces
